terra-base-jupyter/Dockerfile

```diff
@@ -0,0 +1,198 @@
+    && $JUPYTER_USER_HOME/miniconda.sh -b -p $CONDA_ENV_HOME \
+    && rm $JUPYTER_USER_HOME/miniconda.sh
+ENV PATH "${PATH}:${CONDA_ENV_HOME}/bin"
+
+# Set up the path to the user python
+ENV BASE_PYTHON_PATH $CONDA_ENV_HOME/bin/python
+# Tell condo to NOT write bite code (aka these.pyc files)
+ENV PYTHONDONTWRITEBYTECODE=true
+
+###################################################
+# Set up the user to use the conda base environment
+###################################################
+## The user should have full access to the conda base environment, and can use it directly, or
+## create new conda environments on top of it. The important part is that jupyter IS NOT installed
+## in the base environment to provide isolation between the user environment, and the jupyter server
+## to avoid cross-contamination
+COPY conda-environment.yml .
+RUN conda env update --prefix $CONDA_ENV_HOME --file conda-environment.yml --prune \
+    # Remove packages tarballs and python bytecode files from the image
+    && conda clean -afy \
+    && rm conda-environment.yml \
+    # Make sure the JUPYTER_USER is the owner of the folder where
+    # the base conda is installed
+    && chown -R $JUPYTER_USER:users $JUPYTER_USER_HOME
+
+# Add the user base conda environment as a jupyter kernel - this should be the default now
+# This commands activates the conda environment and then calls ipykernel from within
+# to install it as a kernel under the same name
+RUN conda run -p ${CONDA_ENV_HOME} python -m ipykernel install --name=$CONDA_ENV_NAME
+
+# Prep the jupyter terminal to conda init and make sure the base conda environment is 
+# activated and the name is displayed in the terminal prompt
+COPY conda_init.txt .
+RUN cat conda_init.txt >> $JUPYTER_USER_HOME/.bashrc && \
+    printf "\nconda activate ${CONDA_ENV_HOME}" >> $JUPYTER_USER_HOME/.bashrc && \
+    conda config --set env_prompt '({name})' && \
+    source $JUPYTER_USER_HOME/.bashrc && \
+    rm conda_init.txt
+
+####################################################
+# Install Jupyter in an isolated virtual environment
+####################################################
+## Virtualenv and POETRY are the prefered tool to create virtual environments and
+## manage dependencies for Terra Devs - poetry docs: https://python-poetry.org/docs/
+ENV POETRY_HOME /opt/poetry
+# Append POETRY_HOME to PATH
+ENV PATH "${PATH}:${POETRY_HOME}/bin"
+COPY poetry.lock .
+COPY pyproject.toml .
+
+ENV JUPYTER_HOME /usr/jupytervenv
+# Add jupyter virtual environmemt to PATH,
+# but make sure to add it at the end so that the 
+# Conda base python takes precedence 
+# (aka the ! operator in iPython shells should NOT access the jupyter virtualenvironment)
+ENV PATH "${PATH}:${JUPYTER_HOME}/bin"
+
+# Install Poetry, set up the virtual environment for jupyter to run and then cleanup / uninstall poetry
+RUN curl -sSL https://install.python-poetry.org | POETRY_HOME=$POETRY_HOME $BASE_PYTHON_PATH \
```

Comment:
One thing that I was thinking about is maybe not using the user python to install the jupyter server, but instead use the system one to provide another layer of isolation